### PR TITLE
PG-592: Treat queries with different parent queries as separate entries

### PR DESCRIPTION
**Pull request by dutow**
PG-592

1. Previously pg_stat_monitor had a `topquery` and `topqueryid` field, but it was only a sample: it showed one of the top queries executing the specific query.

With this change, the same entry executed by two different functions will result in two entries in the statistics table.

2. This also fixes a bug where the content of these field disappeared for every second query executed: previously the update function changed topqueryid to `0` if it was non zero, and changed it to the proper id when it was 0. This resulted in an alternating behavior, where for every second executed query the top query disappeared.

After these changes, the top query is always shown.

3. The previous implementation also leaked dsa memory used to store the parent queries. This is now also fixed.

### CODE REVIEW

File: pg_stat_monitor.c [hunk 1]
```diff
@@ -1900,6 +1911,8 @@ pgsm_store(pgsmEntry * entry)
 		shared_hash_entry->pgsm_query_id = entry->pgsm_query_id;
 		shared_hash_entry->encoding = entry->encoding;
 		shared_hash_entry->counters.info.cmd_type = entry->counters.info.cmd_type;
+		shared_hash_entry->counters.info.cmd_type = entry->counters.info.cmd_type;
```
Comment (codeforall) [inline pg_stat_monitor.c:1914]: The statement is redundant

File: pg_stat_monitor.c [hunk 2]
```diff
@@ -20,6 +20,7 @@
 #include "nodes/pg_list.h"
 #include "utils/guc.h"
 #include <regex.h>
+#include <assert.h>
```
Comment (EngineeredVirus) [inline pg_stat_monitor.c:23]: This is not needed. See my comments below.

File: pg_stat_monitor.c [hunk 3]
```diff
@@ -1472,11 +1473,7 @@ pgsm_update_entry(pgsmEntry * entry,
 			}
 			else
 			{
-				if(e->counters.info.parent_query != InvalidDsaPointer) {
-					dsa_area   *query_dsa_area = get_dsa_area_for_query_text();
-					dsa_free(query_dsa_area, e->counters.info.parent_query);
-					e->counters.info.parent_query = InvalidDsaPointer;
-				}
+				assert(e->counters.info.parent_query == InvalidDsaPointer);
```
Comment (EngineeredVirus) [inline pg_stat_monitor.c:1476]: Better to change this statement with:

Assert(DsaPointerIsValid(e->counters.info.parent_query) == false);

File: pg_stat_monitor.c [hunk 4]
```diff
@@ -1438,9 +1439,9 @@ pgsm_update_entry(pgsmEntry * entry,
 			e->counters.info.num_relations = num_relations;
 			_snprintf2(e->counters.info.relations, relations, num_relations, REL_LEN);
 
-			if (exec_nested_level > 0 && e->key.parentid != 0 && pgsm_track == PGSM_TRACK_ALL)
+			if (exec_nested_level > 0 && exec_nested_level < max_stack_depth && e->key.parentid != 0 && pgsm_track == PGSM_TRACK_ALL)
 			{
-				if (e->counters.info.parent_query == InvalidDsaPointer && exec_nested_level >= 0 && exec_nested_level < max_stack_depth)
+				if (e->counters.info.parent_query == InvalidDsaPointer)
```
Comment (EngineeredVirus) [inline pg_stat_monitor.c:1444]: Just a thought:
IIUC, this statement is likely to return false most of the time. So, perhaps, this should be the first condition.

Comment (dutow) [reply pg_stat_monitor.c:1444]: If the conditions above are false, this statement will always return true. (if we don't have a proper parent query, the pointer will be invalid)

File: pg_stat_monitor.c (hunk 3 above)
Comment (dutow) [reply pg_stat_monitor.c:1476]: Changed.

File: pg_stat_monitor.c (hunk 2 above)
Comment (dutow) [reply pg_stat_monitor.c:23]: Fixed.

File: pg_stat_monitor.c (hunk 1 above)
Comment (dutow) [reply pg_stat_monitor.c:1914]: Fixed.

**Review by EngineeredVirus — APPROVED** [review]
LGTM

**Review by codeforall — APPROVED** [review]
Looks good

**Review by darkfronza — APPROVED** [review]
LGTM

**Review by Naeem-Akhter — APPROVED** [review]
LGTM!

**Pull request merged**